src/util/storable_builder.rs

```diff
@@ -0,0 +1,71 @@
+use crate::crypto::chacha20poly1305::ChaCha20Poly1305;
+use crate::types::{EncryptionMetadata, PlaintextBlob, Storable};
+use ::prost::Message;
+use rand::rngs::ThreadRng;
+use rand::RngCore;
+use std::borrow::Borrow;
+use std::io;
+use std::io::{Error, ErrorKind};
+
+pub struct StorableBuilder {
+	data_encryption_key: [u8; 32],
+}
+
+const CHACHA20_CIPHER_NAME: &'static str = "ChaCha20Poly1305";
+
+impl StorableBuilder {
+	pub fn build(&self, input: Vec<u8>, version: i64) -> Storable {
+		let mut rng = ThreadRng::default();
+		let mut nonce = [0u8; 12];
+		rng.fill_bytes(&mut nonce[4..]);
+
+		let mut data_blob = PlaintextBlob { value: input, version }.encode_to_vec();
+
+		let mut cipher = ChaCha20Poly1305::new(&self.data_encryption_key, &nonce, &vec![]);
+		let mut tag = [0u8; 16];
+		cipher.encrypt_inplace(&mut data_blob, &mut tag);
+		Storable {
+			data: data_blob,
+			encryption_metadata: Option::from(EncryptionMetadata {
+				nonce: nonce.to_vec(),
+				tag: tag.to_vec(),
+				cipher_format: CHACHA20_CIPHER_NAME.to_string(),
+			}),
```

Comment:
Is this result essentially immediately serialized and written to disk? From an allocation perspective, it's a bit unfortunate we have an intermediary object requiring a few `Vec`s if the alternative is to use something like `Writeable`.

Reply:
> almost in most cases yes. main constituent of it is data_blob, which is encrypted_in_place and decrypted_in_place, and then used as it is in storable. others are just 96bytes and 16bytes. from allocation perspective, i don't think it will allocate unless we encode it. it will just hold references to it.

Any `Vec` creation other than an empty one results in a heap allocation. So there are two when you use `to_vec` here and another heap allocation for `to_string` on `CHACHA20_CIPHER_NAME`. 

> iiuc, even if it was writeable, it would be serializing vec's one after other into one big VecWriter just before writing when encode() is called.

If we didn't use `Storable`, then we could serialize `nonce,` `tag`, and `CHACHA20_CIPHER_NAME` without creating any new `Vec`s or `String`s . True that it would ultimately put everything into a `Vec`, so at least one allocation is required.

If we used a rust struct implementing `Writeable` instead of a proto, we could avoid the unnecessary allocations.